src/xpcc/architecture/interface/memory.hpp

```diff
@@ -0,0 +1,109 @@
+// coding: utf-8
+/* Copyright (c) 2016, Roboterclub Aachen e.V.
+ * All Rights Reserved.
+ *
+ * The file is part of the xpcc library and is released under the 3-clause BSD
+ * license. See the file `LICENSE` for the full license governing this code.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef XPCC_MEMORY_HPP
+#define XPCC_MEMORY_HPP
+
+#include <stdint.h>
+#include <xpcc/architecture/utils.hpp>
+#include <xpcc/architecture/interface/register.hpp>
+
+/**
+ * @ingroup		interface
+ * @defgroup	memory_traits	Memory traits
+ *
+ * Traits of dynamic memory.
+ *
+ * @see		driver
+ * @author	Niklas Hauser
+ */
+
+namespace xpcc
+{
+
+/// Describes access and type of dynamic memory.
+/// @ingroup memory_traits
+enum class
+MemoryTrait : uint16_t
+{
+	AccessSBus = Bit0,			///< Memory is accessible via System-Bus
+	AccessDBus = Bit1,			///< Memory is accessible via Data-Bus
+	AccessIBus = Bit2,			///< Memory is accessible via Instruction-Bus
+
+	AccessDMA = Bit3,			///< Memory is accessible via DMA
+	AccessDMA2D = Bit4,			///< Memory is accessible via 2D DMA
+
+	// Bit5 - Bit12 are reserved
+
+	TypeCoreCoupled = Bit13,	///< Memory is coupled closely to the core
+	TypeNonVolatile = Bit14,	///< Memory is non-volatile (battery-backed)
+	TypeExternal = Bit15,		///< Memory is external RAM
+};
+
+/// A memory can have multiple traits.
+/// @ingroup memory_traits
+using MemoryTraits = Flags16<MemoryTrait>;
+XPCC_TYPE_FLAGS(MemoryTraits);
+
+/// Fast code memory is accessible by at least the I-Bus and it is core-coupled.
+/// Fallback memory on exhaustion is not core-coupled.
+/// @ingroup memory_traits
+constexpr MemoryTraits MemoryFastCode = (MemoryTrait::AccessIBus | MemoryTrait::TypeCoreCoupled);
+/// Fast data memory is accessible by at least the D-Bus and it is core coupled.
+/// Fallback memory on exhaustion is not core-coupled.
+/// @ingroup memory_traits
+constexpr MemoryTraits MemoryFastData = (MemoryTrait::AccessDBus | MemoryTrait::TypeCoreCoupled);
```

Comment:
A better understanding of the fallback mechanism is required. SBus can service both data and instruction requests, but slower, therefore DBus and IBus should always be preferred.

The current mask is interpreted as the _minimum_ requirements, ie. if we add `SBus` the CCM won't be matched.
On heap exhaustion there should perhaps be a three-step process: `requirement` -> `requirement & ~type` -> `(requirement & ~type) | SBus`?
However, that would incur searching three times in case of complete heap exhaustion.
On the other hand, searching longer is better than an out-of-memory scenario.